core/coreapi/dag.go

```diff
@@ -56,6 +61,244 @@ func (api *dagAPI) Session(ctx context.Context) ipld.NodeGetter {
+	//    ipfs dag import $( ... | ipfs-dagger --stdout=carfifos )
+	//
+	unlocker := api.core.blockstore.PinLock()
+	defer unlocker.Unlock()
+
+	retCh := make(chan importResult, 1)
+	go api.importWorker(ctx, directory, retCh)
+
+	rootsCh := make(chan RootMeta, 1)
+	errCh := make(chan error, 1)
+
+	go func() {
+		defer close(errCh)
+		defer close(rootsCh)
+
+		done := <-retCh
+		if done.err != nil {
+			errCh <- done.err
+			return
+		}
+
+		// It is not guaranteed that a root in a header is actually present in the same ( or any )
+		// .car file. This is the case in version 1, and ideally in further versions too
+		// Accumulate any root CID seen in a header, and supplement its actual node if/when encountered
+		// We will attempt a pin *only* at the end in case all car files were well formed
+		//
+		// The boolean value indicates whether we have encountered the root within the car file's
+		roots := done.roots
+
+		// If we are not pinning roots then just return all the roots
+		if !doPinRoots {
+			for r := range roots {
+				select {
+				case rootsCh <- RootMeta{Cid: r}:
+				case <-ctx.Done():
+					errCh <- ctx.Err()
+					return
+				}
+			}
+			return
+		}
+
+		// opportunistic pinning: try whatever sticks
+		if doPinRoots {
+
+			var failedPins int
+			for c := range roots {
+
+				// We need to re-retrieve a block, convert it to ipld, and feed it
+				// to the Pinning interface, sigh...
+				//
+				// If we didn't have the problem of inability to take multiple pinlocks,
+				// we could use the api directly like so (though internally it does the same):
+				//
+				// // not ideal, but the pinning api takes only paths :(
+				// rp := path.NewResolvedPath(
+				// 	ipfspath.FromCid(c),
+				// 	c,
+				// 	c,
+				// 	"",
+				// )
+				//
+				// if err := api.Pin().Add(req.Context, rp, options.Pin.Recursive(true)); err != nil {
+
+				ret := RootMeta{Cid: c}
+
+				if block, err := api.core.blockstore.Get(c); err != nil {
+					ret.PinErrorMsg = err.Error()
+				} else if nd, err := ipld.Decode(block); err != nil {
+					ret.PinErrorMsg = err.Error()
+				} else if err := api.core.pinning.Pin(ctx, nd, true); err != nil {
+					ret.PinErrorMsg = err.Error()
+				} else if err := api.core.pinning.Flush(ctx); err != nil {
+					ret.PinErrorMsg = err.Error()
+				}
+
+				if ret.PinErrorMsg != "" {
+					failedPins++
+				}
+
+				select {
+				case rootsCh <- ret:
+				case <-ctx.Done():
+					errCh <- ctx.Err()
+					return
+				}
+			}
+		}
+	}()
+
+	return rootsCh, errCh
+}
+
+type importResult struct {
+	roots map[cid.Cid]struct{}
+	err   error
+}
+
+func (api *dagAPI) importWorker(ctx context.Context, dir files.Directory, ret chan importResult) {
+	// this is *not* a transaction
+	// it is simply a way to relieve pressure on the blockstore
+	// similar to pinner.Pin/pinner.Flush
+	batch := ipld.NewBatch(ctx, api)
+
+	roots := make(map[cid.Cid]struct{})
+
+	it := dir.Entries()
+	for it.Next() {
```

Comment:
Would this benefit from starting up a number of goroutines (one per entry) and doing the imports concurrently?

Reply:
Almost certainly (up until some limit of worker nodes).

Also, this code does not properly respect contexts. If for some reason the batching datastore underneath does not return an error on `batch.Add(ctx, node)` we won't abort during this process.

Reply:
One could certainly do things here in parallel, but be careful not switching to parallel full-on imports, as this will break a use case hat goes something like:
`ipfs dag import {{named-fifo-pipe-containing-blocks-and-NO-ROOTS-to-pin}} {{named-fifo-pipe-containing-NO-BLOCKS-and-a-list-of-roots-to-pin-based-on-previous-fifo}}`